### PR TITLE
dom/met signature: `make_clamp`

### LINKED ISSUE #722: partial `make_cast`

### PR DESCRIPTION
**Pull request by Shoeboxam**
Closes #722 

Adds `input_domain` and `input_metric` as arguments to `make_clamp`.

* support clamping on sized datasets.
* support other dataset metrics (`InsertDeleteDistance`)
* adds support for partial construction: `prev_trans >> partial_clamp(bounds)`

### CODE REVIEW

**Comment by Shoeboxam** [comment]
Current dependencies on/for this PR:
* main
  * **PR #648** <a href="https://app.graphite.dev/github/pr/opendp/opendp/648" target="_blank">Graphite</a> 
    * **PR #618** <a href="https://app.graphite.dev/github/pr/opendp/opendp/618" target="_blank">Graphite</a> 
      * **PR #674** <a href="https://app.graphite.dev/github/pr/opendp/opendp/674" target="_blank">Graphite</a> 
        * **PR #675** <a href="https://app.graphite.dev/github/pr/opendp/opendp/675" target="_blank">Graphite</a> 
          * **PR #676** <a href="https://app.graphite.dev/github/pr/opendp/opendp/676" target="_blank">Graphite</a> 
            * **PR #678** <a href="https://app.graphite.dev/github/pr/opendp/opendp/678" target="_blank">Graphite</a> 
              * **PR #682** <a href="https://app.graphite.dev/github/pr/opendp/opendp/682" target="_blank">Graphite</a> 
                * **PR #684** <a href="https://app.graphite.dev/github/pr/opendp/opendp/684" target="_blank">Graphite</a> 
      * **PR #645** <a href="https://app.graphite.dev/github/pr/opendp/opendp/645" target="_blank">Graphite</a> 
        * **PR #604** <a href="https://app.graphite.dev/github/pr/opendp/opendp/604" target="_blank">Graphite</a> 
          * **PR #687** <a href="https://app.graphite.dev/github/pr/opendp/opendp/687" target="_blank">Graphite</a> 
            * **PR #688** <a href="https://app.graphite.dev/github/pr/opendp/opendp/688" target="_blank">Graphite</a> 
              * **PR #512** <a href="https://app.graphite.dev/github/pr/opendp/opendp/512" target="_blank">Graphite</a> 
              * **PR #689** <a href="https://app.graphite.dev/github/pr/opendp/opendp/689" target="_blank">Graphite</a> 
                * **PR #690** <a href="https://app.graphite.dev/github/pr/opendp/opendp/690" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/opendp/opendp/690?utm_source=stack-comment).

**Comment by Shoeboxam** [comment]
Current dependencies on/for this PR:
* main
  * **PR #648** <a href="https://app.graphite.dev/github/pr/opendp/opendp/648" target="_blank">Graphite</a> 
    * **PR #618** <a href="https://app.graphite.dev/github/pr/opendp/opendp/618" target="_blank">Graphite</a> 
      * **PR #674** <a href="https://app.graphite.dev/github/pr/opendp/opendp/674" target="_blank">Graphite</a> 
        * **PR #675** <a href="https://app.graphite.dev/github/pr/opendp/opendp/675" target="_blank">Graphite</a> 
          * **PR #676** <a href="https://app.graphite.dev/github/pr/opendp/opendp/676" target="_blank">Graphite</a> 
            * **PR #678** <a href="https://app.graphite.dev/github/pr/opendp/opendp/678" target="_blank">Graphite</a> 
              * **PR #682** <a href="https://app.graphite.dev/github/pr/opendp/opendp/682" target="_blank">Graphite</a> 
                * **PR #684** <a href="https://app.graphite.dev/github/pr/opendp/opendp/684" target="_blank">Graphite</a> 
      * **PR #645** <a href="https://app.graphite.dev/github/pr/opendp/opendp/645" target="_blank">Graphite</a> 
        * **PR #604** <a href="https://app.graphite.dev/github/pr/opendp/opendp/604" target="_blank">Graphite</a> 
          * **PR #687** <a href="https://app.graphite.dev/github/pr/opendp/opendp/687" target="_blank">Graphite</a> 
            * **PR #688** <a href="https://app.graphite.dev/github/pr/opendp/opendp/688" target="_blank">Graphite</a> 
              * **PR #512** <a href="https://app.graphite.dev/github/pr/opendp/opendp/512" target="_blank">Graphite</a> 
              * **PR #689** <a href="https://app.graphite.dev/github/pr/opendp/opendp/689" target="_blank">Graphite</a> 
                * **PR #690** <a href="https://app.graphite.dev/github/pr/opendp/opendp/690" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/opendp/opendp/690?utm_source=stack-comment).

**Review by andrewvyrros — COMMENTED** [review]
Thanks for the great discussion today, I can understand the path to a streamlined API much better now.

Seeing a lot of code in the partial style for the first time, a few observations:

* The `partial_xxx` prefix is a little distracting, especially when mixed in with other code. I'm concerned that people unfamiliar with partial application will be confused. It's particularly awkward with clamp, where `partial_clamp` sounds like it clamps to a half-open range or something. Not sure what to use instead though. `build_xxx`, `chain_xxx`, `pipe_xxx`, `mk_xxx`?
* The first element of the chain gets stuck with a lot of boilerplate. There are a ton of instances of this:
  ```
    from opendp.domains import atom_domain, vector_domain
    from opendp.metrics import symmetric_distance
    input_domain = vector_domain(atom_domain(T=float))
    input_metric = symmetric_distance()

    t = (input_domain, input_metric) >> make_clamp(...)
  ```

  How about a convenience constructor along these lines:
  ```
    def input_space(DI="VectorDomain<AtomDomain<float>>", MI="SymmetricDistance"):
        // Construct a tuple of (DI, MI)

    t1 = input_space() >> make_clamp(...)
    t2 = input_space(DI=VectorDomain<AtomDomain<int>>) >> make_clamp(...)
  ```

**Comment by Shoeboxam** [comment]
I opened https://github.com/opendp/opendp/pull/749 with a convenience constructor as mentioned above.

File: docs/source/examples/pums-data-analysis.ipynb [hunk 1]
```diff
@@ -373,25 +374,26 @@
...
    "source": [
-    "from opendp.transformations import make_clamp, make_bounded_sum\n",
+    "from opendp.transformations import part_clamp, make_bounded_sum\n",
     "bounded_income_sum = (\n",
     "    income_preprocessor >>\n",
-    "    # Clamp income values\n",
-    "    make_clamp(bounds=income_bounds) >>\n",
-    "    # These bounds must be identical to the clamp bounds, otherwise chaining will fail\n",
+    "    # clamp income values. \n",
+    "    # \"partial_*\" means it takes the output domain and metric from the previous transformation\n",
```
Comment (andrewvyrros) [inline docs/source/examples/pums-data-analysis.ipynb:394]:
```suggestion
    "    # \"part_*\" means it takes the output domain and metric from the previous transformation\n",
```

**Review by andrewvyrros — APPROVED** [review]
One small doc typo, feel free to fix it elsewhere if you don't want to wait for CI to run again.

**Pull request merged**